### PR TITLE
Implementation of the multinomial generalized linear model (softmax link)

### PR DESCRIPTION
**Pull request by jachymb**
## Summary

This is the implementation of the `multinomial_logit_glm_lpmf`, including the OpenCL variant. A solution to #3149

It computes, in an efficient vectorized way, exactly what one would expect:

$\mathrm{multinomialLogitGLM} (y|x, \alpha, \beta) = \prod_n \mathrm{multinomial}(y_n \mid \mathrm{softmax}(x_n \beta + \alpha_n))$ 

I was trying to closely follow patterns in related functions like the `binomial_logit_glm_lmpf` and `categorical_logit_glm_lpmf` which this is a generalization of both. 

The typing:

- observations `y` are a 2D `array [,] int`  non-negative (rows=instances, columns=outcome classes)
- thus the predictors `x` are `matrix`(rows=instances, columns=features)
- the coefficients `beta` are a `matrix` (rows=features, columns=outcome classes)
- then the intercept `alpha` may be either a `matrix` (same shape as `y`) or a `row_vector` (length=outcome classes). Now this may seem a little irregular at first, but it makes sense when you really think about it. Other GLMs offer either a `vector` or `real` for the intercept. In the multivariate case, the (univariate) `vector` corresponds to the `matrix` and the scalar to be copied for each instance is more naturally represented here as a `row_vector` rather than a column. (It can be seen as an extra row of `beta` corresponding to a constant feature, also matching the shape of `y` and in consequence, the implementation algebra is just more natural this way.) However this may be argued to be a little inconsistent with `categorical_logit_glm_lpmf` so it's maybe up to a debate, but then the categorical is not implemented as multivariate even though it also uses beta matrix.

I chose to use "logit" in the name instead of "softmax" (even though it really is softmax) to be consistent with existing functions like `multinomial_logit`.

I am new to this, so if there are any improvements to be made, I will gladly listen. I tried to use efficient vectorized approach and utilize existing functions where applicable. Note however this uses its own softmax implementation, which is necessary for efficient vectorization and in my understanding can't be done by the existing library implementation the way it's needed here. Maybe if my other PR #3313 is approved, this could be made a few LOC shorter with a call to that. 

As a personal note, this is a function I want available in Stan for my actual work. :) 

## Tests

Standard testing of the new functions introduced, both for the prim and cl variants.

## Side Effects

None.

## Release notes

Add `multinomial_logit_glm_lpmf`, including OpenCL support.

## AI use disclosure

I used claude code w/ Sonnet to help with the work, but I critically reviewed/edited every single line, striving to match the general code quality and patterns within the library.

## Checklist

- [x] Copyright holder: me, jachymb@gmail.com
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing
    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Coding-Style-and-Idioms) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by jachymb** [comment]
@drezap you said on the forum you could review this, that would be welcome :)

**Comment by WardBrian** [comment]
@lingium may also be a good reviewer, they've recently added several PMFs to Stan

**Comment by stemangiola** [comment]
Tagging @gaoyingnan-academic as we are trying to implement a similar model for compositional data.

**Comment by jachymb** [comment]
update: Added a missing mix test which made me realize there were some minor issues which I also fixed but broke the build temporarily. All works now again.

**Comment by SteveBronder** [comment]
fyi going through this now. Most of the changes I have are very tiny so I'll probably just push up a commit to fix then merge. Everything in general looks good though!

**Comment by SteveBronder** [comment]
Thanks @jachymb I made a few small changes, but otherwise this is good and we will merge after tests pass!

**Review by SteveBronder — DISMISSED** [review]

**Comment by drezap** [comment]
Response, Steve? You are paid.

Best,


Andre Zapico
linkedin.com/in/andre-zapico
gitub.com/drezap


ME Information and Communication Engineering
University of Electronic Science and Technology of China

Consultant, Owner
likely llc
likelyllc.com

Stan Developer
mc-stan.org

BS Mathematical Sciences: Probabilistic Methods
BS Statistics
University of Michigan, Ann Arbor 2017

**Comment by jachymb** [comment]
Hi, can we merge this finally? :)

**Comment by SteveBronder** [comment]
@jachymb yes sorry this involved some math I needed to get used to and has some weird edge cases I am investigating
